src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/MethodToStateMachineRewriter.cs

```diff
@@ -136,6 +138,18 @@ internal abstract class MethodToStateMachineRewriter : MethodToClassRewriter
             {
                 this.proxies.Add(proxy.Key, proxy.Value);
             }
+
+            // create cache local for reference type "this" in Release
+            var thisParameter = originalMethod.ThisParameter;
+            CapturedSymbolReplacement thisProxy;
+            if ((object)thisParameter != null && 
+                thisParameter.Type.IsReferenceType &&
+                proxies.TryGetValue(thisParameter, out thisProxy) &&
+                F.Compilation.Options.OptimizationLevel == OptimizationLevel.Release)
```

Comment:
We have existing tests which fail if the optimization does not check for a class.

---
In reply to: [87637057](https://github.com/dotnet/roslyn/pull/14959#discussion_r87637057) [](ancestors = 87637057)